### PR TITLE
Dennis rust

### PR DESCRIPTION
**Pull request by TrueDoctor**
Implement Mathematical structures like Points and implement proto hitboxes

### CODE REVIEW

File: game_server/src/maths.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    pub fn distance(&self) -> f32 {
+        f32::sqrt(self.distance2())
+    }
+
+    pub fn distance2(&self) -> f32 {
+        self.x * self.x + self.y * self.y
+    }
+}
+
+pub struct Box {
```
Comment (nat-rix) [inline game_server/src/maths.rs]: Refactor Box to AABox (or better) [Box is reserved]

File: game_server/src/maths.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    /// the size may not be smaller than zero
+    pub size: Vec2, 
+}
+
+impl std::ops::Add<Vec2> for Box {
+    type Output = Self;
+    fn add(self, other: Vec2) -> Self {
+        Self {
+            pos: self.pos + other,
+            size: self.size + other,
```
Comment (nat-rix) [inline game_server/src/maths.rs]: Add should probably move the Box => the size doesn't change
```suggestion
            size: self.size,
```

File: game_server/src/maths.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    }
+}
+
+impl std::ops::Sub<Vec2> for Box {
+    type Output = Self;
+    fn sub(self, other: Vec2) -> Self {
```
Comment (nat-rix) [inline game_server/src/maths.rs:93]: This function has to be logically rewritten. At the moment it is a clone of add.

File: game_server/src/maths.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    }
+}
+
+pub struct RBox {
+    /// Point 1
+    pub pos: Vec2,
+    /// Point 2
+    pub size: Vec2, 
+    /// Width Attention manhatten distance!!!
+    pub w: f32,
```
Comment (nat-rix) [inline game_server/src/maths.rs:107]: This property should be considered to be renamed to a better readable property

File: game_server/src/maths.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+impl std::ops::Add<Vec2> for RBox {
+    type Output = Self;
+    fn add(self, other: Vec2) -> Self {
+        Self {
+            pos: self.pos + other,
+            size: self.size + other,
```
Comment (nat-rix) [inline game_server/src/maths.rs]: The same issue as AABox.
```suggestion
            size: self.size,
```

File: game_server/src/maths.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            w: self.w,
+        }
+    }
+}
+
+impl std::ops::Sub<Vec2> for RBox {
+    type Output = Self;
+    fn sub(self, other: Vec2) -> Self {
```
Comment (nat-rix) [inline game_server/src/maths.rs:123]: The same issue as AABox. This function has to be logically rewritten.

File: game_server/src/collide.rs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+impl Collide<Vec2> for Box {
+    fn collides(&self, other: &Vec2) -> bool {
+        self.pos < *other && other < &(self.pos + self.size) 
+    }
+}
+
+impl Collide<Box> for Box {
+    fn collides(&self, other: &Self) -> bool {
```
Comment (nat-rix) [inline game_server/src/collide.rs:20]: over-complicated code could be:
1. simplified
2. optimized
see AABB collision in the web.

File: game_server/src/maths.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+impl std::cmp::PartialEq for Vec2 {
+    fn eq(&self, other: &Self) -> bool {
+        f32::abs(self.x - other.x) < 1e-8
+        && f32::abs(self.y - other.y) < 1e-8
+    }
+}
+
+impl std::cmp::Eq for Vec2 {}
+
+impl Vec2 {
```
Comment (nat-rix) [inline game_server/src/maths.rs:58]: Missing a norm function for obtaining the unit vector. This feature is needed in game_server/src/collide.rs -> collide::RBox::collide. Line 35.

File: game_server/src/collide.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        || other.collides(&(self.pos))
+        || other.collides(&(self.pos + Vec2{x: self.size.x, y: 0.0}))
+        || other.collides(&(self.pos + Vec2{x: 0.0, y: self.size.y}))
+        || other.collides(&(self.pos + self.size))
+    }
+}
+
+impl Collide<Vec2> for RBox {
+    fn collides(&self, other: &Vec2) -> bool {
```
Comment (nat-rix) [inline game_server/src/collide.rs:34]: This function should probably be (completely) re-thinked (especially the manhattan-thing xd) (or the structure of RBoxes)

File: game_server/src/collide.rs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        let ddx = other.x-px; 
+        let ddy = other.y-py;
+        let manhattenDistance = ddx + ddy;
+
+        manhattenDistance < self.w 
+    }
+}
+
+impl Collide<Box> for RBox {
+    fn collides(&self, other: &Box) -> bool {
```
Comment (nat-rix) [inline game_server/src/collide.rs:58]: the (in my opinion not) discussion-worthy thing:
are two **touching** boxes =**colliding**?

File: game_server/src/webhogg_game.rs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+use crate::maths::Vec2;
+
+pub struct WebhoggPlayer {
+    pos: Vec2,
+}
+
+pub struct WebhoggGame {
```
Comment (nat-rix) [inline game_server/src/webhogg_game.rs:7]: Does a WebhoggGame really always have exactly two WebhoggPlayers (also at startup)? Don't forget, Rust does not have null-types. In that case, Options would be the soulution. But of course I'm not aware of your plans for this structure.
But in general: you want the WebhoggGame and the WebhoggPlayer having a constructor function.

File: game_server/src/webhogg_player.rs [hunk 12]
```diff
@@ -0,0 +1,3 @@
+pub struct WebhoggPlayer {
```
Comment (nat-rix) [inline game_server/src/webhogg_player.rs:1]: This declaration is a duplicate of game_server/src/webhogg_game.rs::WebhoggPlayer.
However, this script is not registered into main.rs mod files xd.

**Review by nat-rix — CHANGES_REQUESTED** [review]
Some minor and some less major issues have to be fixed.

File: game_server/src/collide.rs (hunk 9 above)
Comment (TrueDoctor) [reply game_server/src/collide.rs:34]: RBoxes are very usefull for simple hitbok modelling

**Pull request merged**